---
title: "specificity_paper_main"
author: "Leroy Bondhus"
date: "8/19/2021"
output: html_document
---


## NOTE: most functions live in GeneSpecificityFuncs package for 
##       this project. Load this package here
## 
```{r set up package of functions used}
library("devtools")
library("roxygen2")
install("./../GeneSpecificityFuncs")
```

```{r temp:move funcs to pacakge}
#actual standard deviations matrix
make_zscore_matrix<- function(dat) {
  zscores <- matrix(NA, nrow = nrow(dat), ncol = ncol(dat))
  means <- rowMeans(dat)
  sds <- as.numeric(rep(NA,length(means)))
  which <- which(means != 0)
  sds[which] <- apply(dat[which,],1,sd)
  for(j in 1:ncol(dat)){zscores[,j] <- (dat[,j] - means)/sds  }
  return(zscores)
}

make_dot_product_matrix <- function(dat) {
  dot_product_matrix <- matrix(0, nrow = nrow(dat), ncol = ncol(dat))
  for(i in 1:nrow(dat)){
    for(j in 1:nrow(dat)){
      dot_product_matrix[i,j] <- sum(dat[,i] * dat[,j])
    }
  }
  return(dot_product_matrix)
}

actu_sdev <-replace(actu_sdev, is.na(actu_sdev), 0)
n <- matrix(0, nrow = col_num, ncol = col_num)
dot_product_matrix <- matrix(0, nrow = col_num, ncol = col_num)
for(G1 in 1:col_num) {
  for(G2 in 1:col_num)  {
   n <- actu_sdev[,G1]*actu_sdev[,G2]
   dot_product_matrix [G2,G1]<- sum(n)
}
}
diag_val <- diag(dot_product_matrix )
diag_matrix <- data.matrix(diag_val)
simi_matrix<- matrix(0, nrow = col_num, ncol = col_num)
for(G1 in 1:col_num) {
  for(G2 in 1:col_num)  {
 simi_matrix[G1,G2]<-  dot_product_matrix [G1,G2]/(max(diag_matrix [G1],diag_matrix [G2]))
}
}
colnames(simi_matrix)<-colnames(data)
rownames(simi_matrix)<-colnames(data)
View(simi_matrix)


}

```

## load gtex here
```{r import dataset}
## import gtex medians data
temp <- tempfile()
download.file("https://storage.googleapis.com/gtex_analysis_v8/rna_seq_data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct.gz",temp)
gtex <- read.table( temp, skip=2, header = TRUE, sep = "\t")
unlink(temp); rm(temp)



library(biomaRt)
## import ensembl gene data
ensembl = useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl", GRCh=37)
genes <- getBM(attributes=c('chromosome_name','start_position','end_position','hgnc_symbol', 'ensembl_gene_id','gene_biotype'),
                 filters = list('biotype'='protein_coding'),
                 mart = ensembl, useCache = F) 
genes <- genes[which(is.element(genes$chromosome_name, c(1:22, "X", "Y", "MT")) & genes$hgnc_symbol != "" ) ,]



```

## clean gtex dataset here
```{r clean dataset}
## show mitochondrial genes drive a large part of sample similarity
## Note sum of medians in gtex not quite 1M - likely artifact of taking medians
barplot(colSums(gtex[,3:ncol(gtex)]))

#### Supplemental Figure (Fraction of TPM from chr==M )


#### Supplemental Figure (heatmap cluster of samples )

## remove mitochondrial contribution
```


```{r measure sample similarity}
### to do: justify method for measuring sample similarity - may require comparison or refs

### create: sim_mat # sample similarity matrix
```


```{r hierarchical clustering on sample similarity}
### to do: justify method for hierarchical clustering - may require comparison or refs

### create: sim_tree # sample similarity tree
```


```{r calculate sample weights}
### uses Equation 1. 

### create: sample_weights
```


```{r calculate specificity}
### uses Equation 2.

### creates: spec_mat # specificity matrix
```





```{r validation with variable sample set}


```



